ios/adapters/epson/NetPrinterEpsonAdapter.m

```diff
@@ -1,112 +1,160 @@
 
-    } @catch (NSException *exception) {
+  _retryAttempts = 0;
+  _finishedAsyncCall = NO;
 
-          [printer endTransaction];
-          int failResult=[printer disconnect];
-          [printer clearCommandBuffer];
-          [printer setReceiveEventDelegate:nil];
-          errorCallback(@[[NSString stringWithFormat:@"%@ and disconnect code %i",exception.reason,failResult]]);
-    }
-}
+  Epos2Printer *printer =
+      [[Epos2Printer alloc] initWithPrinterSeries:modelNumber
+                                             lang:EPOS2_MODEL_ANK];
+  NSString *printerLock;
 
-- (void) connectAndSendAux:(NSString *)host printRawData:(NSString *)text success:(RCTResponseSenderBlock)successCallback fail:(RCTResponseSenderBlock)errorCallback {
-        NSString* target = [NSString stringWithFormat:@"TCP:%@", host];
-        int result = EPOS2_SUCCESS;
-        result = [printer connect:target timeout:5000];
-        if (result != EPOS2_SUCCESS) {
-            [NSException raise:@"Invalid connection" format:@"Can't connect to printer %@, ERROR code: %i", host,result];
-        }
-
-        NSData* payload = [NSData dataWithBase64EncodedString:text];
+  @synchronized(printerDictionary) {
+    if (printerDictionary == nil) {
+      printerDictionary = [[NSMutableDictionary alloc] init];
+    }
+  }
 
-        [printer addCommand:payload];
-        result = [printer sendData:5000];
+  @synchronized(printerDictionary) {
+    printerLock = [printerDictionary objectForKey:host];
+    if (printerLock == nil) {
+      printerLock = [NSString stringWithFormat:@"%@", host];
+      [printerDictionary setObject:printerLock forKey:host];
+    }
+  }
 
-        if (result != EPOS2_SUCCESS) {
-            [NSException raise:@"Print failed" format:@"Error occurred while printing"];
+  @synchronized(printerLock) {
+    @try {
+      __weak NetPrinterEpsonAdapter *weakSelf = self;
+      [printer setReceiveEventDelegate:weakSelf];
+      do {
+        @try {
+          _retryAttempts++;
+          [self connectAndSendAux:host
+                     printRawData:text
+                          success:successCallback
+                             fail:errorCallback
+                       printerObj:printer];
+          _retryAttempts = 3;
+        } @catch (NSException *exception) {
+          if (_retryAttempts >= 3) {
+            @throw exception;
+          }
+          [printer endTransaction];
+          [printer disconnect];
+          [printer clearCommandBuffer];
+          [NSThread sleepForTimeInterval:(3.0f * _retryAttempts)];
         }
+      } while (_retryAttempts < 3);
+
+      long delayInSeconds = 3;
+      long startTime = [[NSDate date] timeIntervalSince1970];
+      long currTime = [[NSDate date] timeIntervalSince1970];
+      while (!_finishedAsyncCall && startTime + delayInSeconds > currTime) {
+        [NSThread sleepForTimeInterval:0.5];
+        currTime = [[NSDate date] timeIntervalSince1970];
+      }
```

Comment:
Good, there is a sleep in this loop to ensure the checks is not going nuts CPU-wise